### PR TITLE
List programs

### PR DESCRIPTION
**Pull request by balopat**
Exposes the listing of quantum programs on the Engine API.

One half of #3302.

### CODE REVIEW

File: cirq/google/engine/engine_client_test.py [hunk 1]
```diff
@@ -94,6 +94,54 @@ def test_get_program(client_constructor):
...
+        (datetime.datetime(2020, 9, 1, 0, 0,
+                           0), None, None, 'create_time >= 1598932800'),
+        (None, datetime.date(2020, 10, 1), None, 'create_time <= 2020-10-01'),
+        (datetime.date(2020, 9, 1), datetime.datetime(
+            2020, 9, 1, 0, 0, 10), None, 'create_time >= 2020-09-01 AND '
+         'create_time <= 1598932810'),
+        (None, None, {
+            'color': 'red',
+            'shape': '*'
+        }, 'labels.color:red AND labels.shape:*'),
```
Comment (wcourtney) [inline cirq/google/engine/engine_client_test.py:129]: Consider one with all filter types just for giggles?

**Review by wcourtney — APPROVED** [review]
Mostly nit-picky comments, so approving as-is.

File: cirq/google/engine/engine.py [hunk 2]
```diff
@@ -482,6 +482,43 @@ def get_program(self, program_id: str) -> engine_program.EngineProgram:
...
+                          Union[datetime.datetime, datetime.date]] = None,
+                      created_after: Optional[
+                          Union[datetime.datetime, datetime.date]] = None,
+                      has_labels: Optional[Dict[str, str]] = None
+                     ) -> List[engine_program.EngineProgram]:
+        """Returns a list of previously executed quantum programs.
+
+        Args:
+            created_after: filter programs that were created after this date or
+             time. Note that datetimes are converted to UTC.
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/google/engine/engine.py:496]: I can skip it.

File: cirq/google/engine/engine.py [hunk 3]
```diff
@@ -482,6 +482,43 @@ def get_program(self, program_id: str) -> engine_program.EngineProgram:
...
+            created_before: filter programs that were created after this date or
+             time. Note that datetimes are converted to UTC.
+            has_labels: filter programs that have labels on them specified by
+                this dict. If the value is set to '*', filters having the label
+                regardless of the label value will be filtered. For example, to
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/google/engine/engine.py:501]: I was wondering about that for a second - thanks. I'll rephrase to "retrieve programs that have labels.."

File: cirq/google/engine/engine_client.py [hunk 4]
```diff
@@ -199,6 +200,55 @@ def get_program(self, project_id: str, program_id: str,
...
+            val = _to_filter_date_or_time('created_after', created_after)
+            filters.append(f"create_time >= {val}")
+        if created_before:
+            val = _to_filter_date_or_time('created_before', created_before)
+            filters.append(f"create_time <= {val}")
+        if has_labels:
+            for (k, v) in has_labels.items():
+                filters.append(f"labels.{k}:{v}")
+        return self._make_request(
+            lambda: self.grpc_client.list_quantum_programs(
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/google/engine/engine_client.py:249]: What is the default page size? It seems to retrieve 241 programs without a problem.

**Pull request merged**